main/src/ca/uwaterloo/flix/language/phase/Inliner1.scala

```diff
@@ -0,0 +1,986 @@
+            val e1 = visitExp(exp1, varSubst0, subst0, inScopeSet0, context0)
+            // Case 4:
+            // If `e1` is trivial and pure, then it is safe to inline.
+            // Code size and runtime are not impacted, because only trivial expressions are inlined
+            val wantToPostInline = isTrivialAndPure(e1, exp1.eff) && occur != DontInline
+            if (wantToPostInline) {
+              // If `e1` is to be inlined:
+              // Add map `sym` to `e1` and return `e2` without constructing the let expression.
+              val subst1 = subst0 + (freshVarSym -> SubstRange.DoneExp(e1))
+              visitExp(exp2, varSubst1, subst1, inScopeSet0, context0)
+            } else {
+              // Case 5:
+              // If none of the previous cases pass, `sym` is not inlined. Return a let expression with the visited expressions
+              // Code size and runtime are not impacted
+              val inScopeSet1 = inScopeSet0 + (freshVarSym -> Definition.LetBound(e1, occur))
+              val e2 = visitExp(exp2, varSubst1, subst0, inScopeSet1, context0)
+              MonoAst.Expr.Let(freshVarSym, e1, e2, tpe, eff, loc)
+            }
+          }
+        }
+
+      case OccurrenceAst1.Expr.LocalDef(sym, fparams, exp1, exp2, tpe, eff, occur, loc) =>
+        // TODO: Treat like let-binding above, except it is never trivial so some checks can be omitted
+        if (isDead(occur)) { // Probably never happens
+          visit(exp2)
+        } else {
+          val freshVarSym = Symbol.freshVarSym(sym)
+          val varSubst1 = varSubst0 + (sym -> freshVarSym)
+          val e2 = visitExp(exp2, varSubst1, subst0, inScopeSet0, context0)
+          val (fps, varSubsts) = fparams.map(freshFormalParam).unzip
+          val varSubst2 = varSubsts.foldLeft(varSubst1)(_ ++ _)
+          val e1 = visitExp(exp1, varSubst2, subst0, inScopeSet0, context0)
+          MonoAst.Expr.LocalDef(freshVarSym, fps, e1, e2, tpe, eff, loc)
+        }
+
+      case OccurrenceAst1.Expr.Scope(sym, rvar, exp, tpe, eff, loc) =>
+        val freshVarSym = Symbol.freshVarSym(sym)
+        val varSubst1 = varSubst0 + (sym -> freshVarSym)
+        val e = visitExp(exp, varSubst1, subst0, inScopeSet0, context0)
+        MonoAst.Expr.Scope(freshVarSym, rvar, e, tpe, eff, loc)
+
+      case OccurrenceAst1.Expr.IfThenElse(exp1, exp2, exp3, tpe, eff, loc) =>
+        val e1 = visit(exp1)
+        val e2 = visit(exp2)
+        val e3 = visit(exp3)
+        e1 match {
+          case MonoAst.Expr.Cst(Constant.Bool(true), _, _) => e2
+          case MonoAst.Expr.Cst(Constant.Bool(false), _, _) => e3
+          case _ => MonoAst.Expr.IfThenElse(e1, e2, e3, tpe, eff, loc)
+        }
+
+      case OccurrenceAst1.Expr.Stm(exp1, exp2, tpe, eff, loc) =>
+        // Case 1:
+        // If `exp1` is pure, so it has no side effects, then it is safe to remove
+        // Both code size and runtime are reduced
+        if (isPure(exp1.eff) && !isHole(exp1)) {
```

Comment:
I'd prefer if you not special case hole like this since it only works sometimes and just hides general problems. If you want to respect holes you need to put them into the dangerous def category